File: joyhousebot/api/routers/works.py
"""Personal works, immutable versions, sharing, and public presentation."""

from __future__ import annotations

from fastapi import APIRouter, Response

from joyhousebot.api.dependencies import ContainerDep, ContextDep
from joyhousebot.api.schemas import (
    CreateWorkRequest,
    CreateWorkShareRequest,
    CreateWorkVersionRequest,
    GrantWorkCollaboratorRequest,
    UpdateWorkRequest,
)
from joyhousebot.api.work_schemas import (
    CreateWorkHandoffReceiptRequest,
    CreateWorkHandoffRequest,
)

router = APIRouter(tags=["works"])


@router.get("/works")
async def list_works(context: ContextDep, container: ContainerDep):
    return {"items": await container.works.list(context)}


@router.post("/works", status_code=201)
async def create_work(
    body: CreateWorkRequest, context: ContextDep, container: ContainerDep
):
    return await container.works.create(context, body.model_dump())


@router.get("/works/{work_id}")
async def get_work(work_id: str, context: ContextDep, container: ContainerDep):
    return await container.works.get(context, work_id)


@router.patch("/works/{work_id}")
async def update_work(
    work_id: str,
    body: UpdateWorkRequest,
    context: ContextDep,
    container: ContainerDep,
):
    return await container.works.update(
        context, work_id, body.model_dump(exclude_unset=True)
    )


@router.post("/works/{work_id}/versions", status_code=201)
async def create_work_version(
    work_id: str,
    body: CreateWorkVersionRequest,
    context: ContextDep,
    container: ContainerDep,
):
    return await container.works.add_version(context, work_id, body.model_dump())


@router.get("/works/{work_id}/consumers")
async def list_work_consumers(
    work_id: str, context: ContextDep, container: ContainerDep
):
    return {"items": await container.works.list_consumers(context, work_id)}


@router.get("/works/{work_id}/handoffs")
async def list_work_handoffs(
    work_id: str, context: ContextDep, container: ContainerDep
):
    return {"items": await container.works.list_handoffs(context, work_id)}


@router.post("/works/{work_id}/handoffs", status_code=201)
async def create_work_handoff(
    work_id: str,
    body: CreateWorkHandoffRequest,
    context: ContextDep,
    container: ContainerDep,
):
    return await container.works.create_handoff(context, work_id, body.model_dump())


@router.get("/work-handoffs/{handoff_id}/input")
async def get_work_handoff_input(
    handoff_id: str, context: ContextDep, container: ContainerDep
):
    return await container.works.handoff_input(context, handoff_id)


@router.get("/work-handoffs/{handoff_id}/receipts")
async def list_work_handoff_receipts(
    handoff_id: str, context: ContextDep, container: ContainerDep
):
    return {"items": await container.works.list_handoff_receipts(context, handoff_id)}


@router.post("/work-handoffs/{handoff_id}/receipt", status_code=201)
async def create_work_handoff_receipt(
    handoff_id: str,
    body: CreateWorkHandoffReceiptRequest,
    context: ContextDep,
    container: ContainerDep,
):
    return await container.works.add_handoff_receipt(
        context, handoff_id, body.model_dump()
    )


@router.post("/work-handoffs/{handoff_id}/cancel")
async def cancel_work_handoff(
    handoff_id: str, context: ContextDep, container: ContainerDep
):
    return await container.works.cancel_handoff(context, handoff_id)


@router.get("/works/{work_id}/shares")
async def list_work_shares(
    work_id: str, context: ContextDep, container: ContainerDep
):
    return {"items": await container.works.list_shares(context, work_id)}


@router.post("/works/{work_id}/shares", status_code=201)
async def create_work_share(
    work_id: str,
    body: CreateWorkShareRequest,
    context: ContextDep,
    container: ContainerDep,
):
    return await container.works.create_share(context, work_id, body.model_dump())


@router.post("/works/{work_id}/shares/{share_id}/revoke")
async def revoke_work_share(
    work_id: str,
    share_id: str,
    context: ContextDep,
    container: ContainerDep,
):
    return await container.works.revoke_share(context, work_id, share_id)


@router.get("/works/{work_id}/collaborators")
async def list_work_collaborators(
    work_id: str, context: ContextDep, container: ContainerDep
):
    return {"items": await container.works.list_collaborators(context, work_id)}


@router.put("/works/{work_id}/collaborators/{user_id}")
async def grant_work_collaborator(
    work_id: str,
    user_id: str,
    body: GrantWorkCollaboratorRequest,
    context: ContextDep,
    container: ContainerDep,
):
    value = body.model_dump()
    value["user_id"] = user_id
    return await container.works.grant_collaborator(context, work_id, value)


@router.delete("/works/{work_id}/collaborators/{user_id}", status_code=204)
async def revoke_work_collaborator(
    work_id: str,
    user_id: str,
    context: ContextDep,
    container: ContainerDep,
):
    await container.works.revoke_collaborator(context, work_id, user_id)
    return Response(status_code=204)


@router.get("/works/{work_id}/audit")
async def list_work_audit(
    work_id: str, context: ContextDep, container: ContainerDep
):
    return {"items": await container.works.audit(context, work_id)}


@router.get("/public/works/{slug}")
async def get_public_work(slug: str, container: ContainerDep):
    return await container.works.resolve_public(slug=slug)


@router.get("/public/shares/{token}")
async def get_shared_work(token: str, container: ContainerDep):
    return await container.works.resolve_share(token=token)

File: apps/console/src/api/works.ts
import { apiFetch } from './http'
import { getIdentityHeaders } from './identity'

export interface WorkVersion {
  version: number
  media_type: string
  content?: unknown
  uri?: string | null
  content_sha256: string
  change_note: string
  source_run_id?: string
  source_artifact_id?: string
}

export interface Work {
  work_id: string
  owner_user_id: string
  public_slug: string
  title: string
  description: string
  status: 'draft' | 'published' | 'archived'
  visibility: 'private' | 'unlisted' | 'public'
  data_classification: 'public' | 'internal' | 'confidential' | 'restricted'
  current_version: number
  published_version?: number | null
  metadata: Record<string, unknown>
  updated_at: string
  published_at?: string | null
  version?: WorkVersion | null
}

export interface WorkShare {
  share_id: string
  work_id: string
  version: number
  permission: 'view' | 'download'
  status: string
  expires_at?: string | null
  token?: string
  path?: string
}

export interface WorkConsumer {
  installation_id: string
  app_id: string
  app_version: string
  app_name: string
  consumer_id: string
  name: string
  description: string
  purposes: string[]
  media_types: string[]
  input_schema: Record<string, unknown>
}

export interface WorkHandoff {
  handoff_id: string
  work_id: string
  work_version: number
  installation_id: string
  app_id: string
  app_version: string
  consumer_id: string
  purpose: string
  status: 'authorized' | 'accepted' | 'executing' | 'verified' | 'failed' | 'cancelled'
  content_sha256: string
  created_at: string
  updated_at: string
  completed_at?: string | null
}

async function request<T>(path: string, init: RequestInit = {}): Promise<T> {
  const headers = new Headers(init.headers)
  Object.entries(getIdentityHeaders()).forEach(([key, value]) => headers.set(key, value))
  const response = await apiFetch(`/v1${path}`, { ...init, headers })
  const payload = response.status === 204 ? null : await response.json().catch(() => ({}))
  if (!response.ok) throw new Error(payload?.error?.message ?? payload?.detail ?? '作品 API 调用失败')
  return payload as T
}

export async function listWorks(): Promise<Work[]> { return (await request<{ items: Work[] }>('/works')).items }
export async function getWork(id: string): Promise<Work> { return request(`/works/${encodeURIComponent(id)}`) }
export async function createWork(value: Record<string, unknown>): Promise<Work> {
  return request('/works', { method: 'POST', headers: { 'Content-Type': 'application/json', 'Idempotency-Key': crypto.randomUUID() }, body: JSON.stringify(value) })
}
export async function updateWork(id: string, value: Record<string, unknown>): Promise<Work> {
  return request(`/works/${encodeURIComponent(id)}`, { method: 'PATCH', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(value) })
}
export async function createWorkVersion(id: string, value: Record<string, unknown>): Promise<Work> {
  return request(`/works/${encodeURIComponent(id)}/versions`, { method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(value) })
}
export async function listWorkConsumers(id: string): Promise<WorkConsumer[]> { return (await request<{ items: WorkConsumer[] }>(`/works/${encodeURIComponent(id)}/consumers`)).items }
export async function listWorkHandoffs(id: string): Promise<WorkHandoff[]> { return (await request<{ items: WorkHandoff[] }>(`/works/${encodeURIComponent(id)}/handoffs`)).items }
export async function createWorkHandoff(id: string, value: Record<string, unknown>): Promise<WorkHandoff> {
  return request(`/works/${encodeURIComponent(id)}/handoffs`, { method: 'POST', headers: { 'Content-Type': 'application/json', 'Idempotency-Key': crypto.randomUUID() }, body: JSON.stringify(value) })
}
export async function cancelWorkHandoff(id: string): Promise<WorkHandoff> { return request(`/work-handoffs/${encodeURIComponent(id)}/cancel`, { method: 'POST' }) }
export async function listWorkShares(id: string): Promise<WorkShare[]> { return (await request<{ items: WorkShare[] }>(`/works/${encodeURIComponent(id)}/shares`)).items }
export async function createWorkShare(id: string, value: Record<string, unknown>): Promise<WorkShare> {
  return request(`/works/${encodeURIComponent(id)}/shares`, { method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(value) })
}
export async function revokeWorkShare(id: string, shareId: string): Promise<WorkShare> {
  return request(`/works/${encodeURIComponent(id)}/shares/${encodeURIComponent(shareId)}/revoke`, { method: 'POST' })
}
export async function listWorkAudit(id: string): Promise<Array<Record<string, unknown>>> { return (await request<{ items: Array<Record<string, unknown>> }>(`/works/${encodeURIComponent(id)}/audit`)).items }
export async function listWorkCollaborators(id: string): Promise<Array<Record<string, string>>> { return (await request<{ items: Array<Record<string, string>> }>(`/works/${encodeURIComponent(id)}/collaborators`)).items }
export async function grantWorkCollaborator(id: string, userId: string, role: 'viewer' | 'editor'): Promise<Record<string, string>> {
  return request(`/works/${encodeURIComponent(id)}/collaborators/${encodeURIComponent(userId)}`, { method: 'PUT', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify({ role }) })
}
export async function revokeWorkCollaborator(id: string, userId: string): Promise<void> { await request(`/works/${encodeURIComponent(id)}/collaborators/${encodeURIComponent(userId)}`, { method: 'DELETE' }) }
